import prisma from "../lib/prisma/index.js";
import { z } from "zod";
const docs = `
Documentation:

C2M2: The Crosscut Metadata Model (C2M2) was designed to standardize metadata annotations across Common Fund DCCs by the original CFDE Coordination Center. C2M2 datapackage files are usually a zipfile containing a set of TSV files already standardized to a set of known ontologies, along with a single JSON schema.


RNA-seq Standard Pipeline:
The CFDE RNA-seq Partnership aims to harmonize diverse RNA-seq datasets across the CFDE to improve their interoperability and reusability for the broader scientific research community. To this end, the partnership is developing common pipelines for both bulk and single cell RNA-seq data analysis, with the end goal of building a large, readily accessible, and harmonized resource of RNA-seq datasets and processing tools from various DCCs and CFDE projects.


More information about the standardized bulk and single cell RNA-seq pipelines will be available soon.


OpenAI and SmartAPI:
The OpenAPI specification allows for the standardization of application programming interfaces (APIs), and facilitates their interoperability.


The SmartAPI project builds on top of the OpenAPI specifications to maximize the findability, accessibility, interoperability, and reusability (FAIRness) of web-based tools, especially those within the biomedical sphere. Through richer metadata descriptions, SmartAPI enables users to search and use a connected network of tools.


Knowledge Graph Assertions:
The CFDE Data Distillery Partnership aims to integrate data assertions across DCCs into a functional knowledge graph for knowledge query and discovery. The partnership has collected "distilled" data relationships from each DCC to be unified in a knowledge graph model with controlled ontology and vocabulary terms for exploring pre-defined, biologically relevant use cases.
More information about generating knowledge graph assertions will be available soon.


Fair evaluations with FAIRshake:
The FAIRshake toolkit enables manual and automated assessments of the findability, accessibility, interoperability, and reusability (FAIRness) of digital resources. FAIRshake provides community-driven metrics and rubrics for evaluation, and visualizes the results with a characteristic embeddable insignia. The primary goal of FAIRshake is to enable researchers and developers to objectively measure and improve the FAIRness of their tools.


Playbook Metanodes:
The Playbook Workflow Builder (PWB) is a web-based knowledge resolution platform being developed by the CFDE Workflow Playbook Partnership and consisting of a growing network of datasets, semantically annotated API endpoints, and visualization tools from across the CFDE. Users can construct workflows from the individual building blocks, termed "meta nodes", with little effort or technical expertise required.
`

const cfde_info = `
The Common Fund Data Ecosystem (CFDE)  aims to enable broad use of  Common Fund data to accelerate discovery. Common Fund programs generate a wide range of diverse and valuable data sets and knowledge designed to be used by the research community. However, these data and knowledge reside in different locations, and it is challenging to work with multiple digital resources in an accessible and user-friendly way. To help remedy this problem, the CFDE created an online discovery portal (CFDE Portal) in its first phase to help make Common Fund data sets FAIR (Findable, Accessible, Interoperable, and Reusable) and to enable researchers to search across Common Fund data sets to ask scientific and clinical questions from a single access point.
Throughout its first phase, the CFDE Coordinating Center oversaw CFDE activities and worked closely with participating data coordinating centers from other Common Fund programs to build the data ecosystem. As the CFDE enters its second phase, the diversity of expertise needed to coordinate CFDE actives is increasing, and the program is restructuring to accommodate these new needs. In Phase 2, the CFDE will be structured around the following five centers:


The Data Resource Center (DRC) will create a portal to enable users to query and use the data sets and other digital resources from across Common Fund programs


The Knowledge Center (KC) will create a knowledge network to integrate knowledge generated by Common Fund programs


The Cloud Workspace Implementation Center (CWIC) will support cloud computing activities in select cloud workspaces


The Center for Training will coordinate the training efforts of the CFDE consortium to allow any researcher, regardless of experience, to engage with CFDE tools and resources.


The Integration and Coordination Center will support integrating CFDE-related activities between the CFDE Centers and Common Fund data coordinating centers and perform evaluation activities


The CFDE Portal will integrate resource and knowledge portals developed by the DRC and KC, and also serve as a landing page for users to access other CFDE-generated tools, resources, trainings, and guidance. Together, these resources will empower the research community to use Common Fund data sets for novel scientific research that was impossible before. This may include hypothesis generation, discovery, or validation that leads to new insights in health and disease. The CFDE is also supporting pilot projects that engage a broad community of end-users, collect feedback on the utility of Common Fund data resources, and enable novel cross-cutting biological questions to be formulated and addressed.
DCCs:
Knockout Mouse Phenotyping Program (KOMP2)
The Knockout Mouse Phenotyping Program (KOMP2) collaborates with the International Mouse Phenotyping Consortium (IMPC) to knockout and characterize all protein-coding genes in the mouse genome. Knocking out the activity of a gene provides valuable clues about what that gene normally does but creating mice with gene knockouts is a time-consuming and difficult process, and often is done by individual research labs with varied approaches.


NIH Integrative Human Microbiome Project (iHMP)
The NIH Common Fund Human Microbiome Project (HMP) was established in 2008, with the mission of generating resources that would enable the comprehensive characterization of the human microbiome and analysis of its role in human health and disease. The HMP characterized the microbial communities from 300 healthy individuals, across several different sites on the human body: nasal passages, oral cavity, skin, gastrointestinal tract, and urogenital tract. 16S rRNA sequencing was performed to characterize the complexity of microbial communities at each body sites, and to begin to ask investigate whether there is a core healthy microbiome. More information, and over 14 terabytes of data, are available on the HMP1 side of this website.
Extracellular RNA Communication (ExRNA)
The Extracellular RNA Communication program is exploring the biology of extracellular RNA (exRNA). Once thought to exist only within cells, in a paradigm shift for science, RNA is now known to be exported from cells as “extracellular RNA” and to play a role in cell-to-cell communication. Since its founding, the program has established data standards, a data portal, and tools and reagents available to the scientific community. The program cataloged exRNA molecules found in human biofluids like plasma, saliva, and urine from over 2000 donors. It also identified potential exRNA biomarkers for nearly 30 diseases, including cardiovascular disease, pregnancy complications, glaucoma, diabetes, and multiple types of cancer.


Glycoscience (Glycoscience)
The goal of the Common Fund’s Glycoscience program is to create new resources, tools, and method to make the study of glycans (sugars) more accessible to the broader research community. Glycans play a critical role in nearly all aspects of biology, ranging from how our bodies recognize and fight viruses and bacteria to how proteins are moved throughout our cells to perform different tasks. Despite being an important part to nearly all biological functions, the study of glycans and the roles they play is limited due to the complexity surrounding the chemistry to make, study, and sequence all the different glycans in our body. To combat this, the Glycoscience program has developed a variety of tools, resources, and information for researchers unfamiliar with glycans to incorporate them into their studies. This includes techniques for identifying especially difficult glycans, probes to study glycans in bacterial cell walls, standards for glycan chemical synthesis, methods for high-throughput glycan studies, and glycoscience educational materials.


Undiagnosed Diseases Network (UDN)
The NIH Undiagnosed Diseases Network (UDN) is a research study to improve the level of diagnosis of rare and undiagnosed conditions. In the United States, it has been estimated that approximately 25 million Americans suffer from a rare disorder. The UDN established a nationwide network of clinicians and researchers who use both basic and clinical research to uncover the underlying disease mechanisms associated with these conditions. Since its launch, the UDN has provided over 600 diagnoses and that number continues to grow. The network has discovered hundreds of novel disease-associated genes and genomic variants, including new diseases and syndromes; and built an international reputation for establishing exemplary clinical practices, standards, and pipelines for genomics-based diagnoses.


Gabriella Miller Kids First Pediatric Research (Kids First)
The goal of the Gabriella Miller Kids First Pediatric Research Program (Kids First) is to help researchers uncover new insights into the biology of childhood cancer and structural birth defects, including the discovery of shared genetic pathways between these disorders. Children with birth defects have an increased risk of developing childhood cancer. This suggests there are shared genetic pathways underlying some types of childhood cancer and structural birth defects. Kids First is achieving this goal through two initiatives: 1) identifying children with childhood cancer and structural birth defects and their families for whole genome sequencing performed by the Kids First sequencing centers 2) developing the Gabriella Miller Kids First Data Resource, a large-scale database of clinical and genetic data from patients with childhood cancers and structural birth defects and their families. Analyzing genetic sequence data from children with childhood cancer and structural birth defects together may lead to the discovery of new genetic pathways.


Genotype Tissue Expression (GTEx)
The Common Fund's Genotype-Tissue Expression (GTEx) Program established a data resource and tissue bank to study the relationship between genetic variants (inherited changes in DNA sequence) and gene expression (how genes are turned on and off) in multiple human tissues and across individuals. GTEx also increased our understanding of how gene expression varies between male and female. (Retrieved from the NIH Common Fund site)


Molecular Transducers of Physical Activity Consortium (MoTrPAC)
The Molecular Transducers of Physical Activity Consortium (MoTrPAC) aims to uncover, at the molecular level, how exercise improves and maintains the health of the body’s tissues and organs. Tracking exercise’s impact on biological molecules through time will help MoTrPAC researchers create a map of molecular changes in the body. Studies in three human tissues will be complemented by more extensive studies in an animal model, allowing many additional tissue types to be explored and substantially increasing the impact of the consortium. Researchers will use the molecular map to better understand how exercise affects people of different ages, sexes, body compositions, and fitness levels. In the future, it may also enable clinicians to make more specific exercise recommendations to patients when using exercise as an intervention to improve health. 


Bridge to Artificial Intelligence (Bridge2AI)
The NIH Common Fund’s Bridge to Artificial Intelligence (Bridge2AI) program will propel biomedical research forward by setting the stage for widespread adoption of artificial intelligence (AI) that tackles complex biomedical challenges beyond human intuition. A key step in this process is generating new “flagship” data sets and best practices for machine learning analysis. Machine learning is a type of Artificial Intelligence (AI) that provides machines (like computers) the ability to automatically learn and improve from experience, without being explicitly programmed. The biomedical research community generates a wealth of data, but most of these data are not suitable for machine learning because they are incomplete. By bringing technological and biomedical experts together with social scientists and humanists, the Bridge2AI program will help bring solutions to this deficit. 


Metabolomics (Metabolomics Workbench)
The goal of the Common Fund’s Metabolomics program is to inform basic, translational, and clinical research. Metabolomics is the scientific study of the chemical reactions that occur in organisms, cells, or tissues. Each reaction produces small chemicals called metabolites, which play critical roles in keeping our cells healthy and functioning properly. As these different chemical reactions and the metabolites they produce are unique to every individual, by improving metabolomics methods and making them more accessible to different researchers it may allow for more personalized diagnosis of disease and treatment methods moving forward. 


Illuminating the Druggable Genome (IDG)
To improve our scientific understanding of understudied members of these three protein families, the National Institutes of Health (NIH) Common Fund launched the Illuminating the Druggable Genome (IDG) Program in 2014. The overall goal of the IDG Program is to catalyze research in areas of biology that are currently understudied but that have high potential to impact human health.


The Cellular Senescence Network (SenNet)
The Common Fund’s Cellular Senescence Network (SenNet) Program was established to comprehensively identify and characterize the differences in senescent cells across the body, across various states of human health, and across the lifespan. SenNet will provide publicly accessible atlases of senescent cells, the differences among them, and the molecules they secrete, using data collected from multiple human and model organism tissues. To identify and characterize these rare cells, SenNet will develop innovative tools and technologies that build upon previous advances in single cell analysis, such as those from the Common Fund’s Human Biomolecular Atlas Program and Single Cell Analysis Program. Lastly, SenNet aims to unite cellular senescence researchers by developing common terms and classifications for senescent cells.


4D Nucleome (4DN)
The goal of the Common Fund’s 4D Nucleome (4DN) program is to study the three-dimensional organization of the nucleus in space and time (the 4th dimension). The nucleus of a cell contains DNA, the genetic “blueprint” that encodes all of the genes a living organism uses to produce proteins needed to carry out life-sustaining cellular functions. Scientists know that how the information in the nucleus organized, stored, and unpackaged are all important to basic human health and we are only starting to learn how changes in this organization can lead to the development of different diseases, such as cancer or our response to infectious agents like viruses.


Human BioMolecular Atlas Program (HuBMAP)
The goal of the Human BioMolecular Atlas Program (HuBMAP) is to develop an open and global platform to map healthy cells in the human body. In humans, the proper functioning of organs and tissues is dependent on the interaction, spatial organization, and specialization of all our cells. Scientists estimate there are 37 trillion cells in an adult human body, so determining the function and relationship among these cells is a monumental undertaking. Using the latest molecular and cellular biology technologies, HuBMAP researchers are studying the connections that cells have with each other throughout the body. HuBMAP will build the framework necessary to construct the tools, resources, and cell atlases needed to determine how the relationships between cells can affect the health of an individual.


Human Heredity & Health in Africa (H3Africa)
H3Africa Consortium facilities modern research approaches to study genomic and environmental determinates of common diseases with the goal of improving the health of African populations. The initiative consists of 48 African projects that include population-based genomic studies of common, non-communicable disorders such as heart and renal disease, as well as communicable diseases such as tuberculosis.


Library of Integrated Network-based Cellular Signatures (LINCS)
The Common Fund’s Library of Integrated Network-based Cellular Signatures (LINCS) program developed a library that catalogs changes that occur when different types of cells are exposed to a variety of agents that disrupt normal cellular functions, called “perturbagens.” This library integrated data across cell types, perturbagens, and measurements of cellular responses for use by scientists to design and conduct experiments that can help us better understand normal and diseased cellular states and how to alter them. 


Stimulating Peripheral Activity to Relieve Conditions (SPARC)
The Common Fund’s Stimulating Peripheral Activity to Relieve Conditions (SPARC) program accelerates development of therapeutic devices that modulate electrical activity in nerves to improve organ function. This therapeutic strategy, also known as “bioelectronic medicine,” could offer new treatment options for diverse diseases and conditions such as hypertension, heart failure, gastrointestinal disorders, and more. 


Acute to Chronic Pain Signatures (A2CPS)
The goal of the Acute to Chronic Pain Signatures (A2CPS) program is to develop a set of objective biomarkers that provide “signatures” to predict if chronic pain is likely to develop or be resolved after acute pain, like an injury or after a surgery. These signatures are greatly needed because prevention of chronic pain is a major challenge in pain management. For most people, acute pain resolves as the injury that caused it heals. Yet in many other people, pain from an injury, surgery, or disease persists lasts for years or even throughout life becoming a chronic condition. This high prevalence of chronic pain has in part contributed to the current opioid epidemic in the US. A signature that could be identified before the transition from acute to chronic pain could help accelerate therapy development and ultimately guide pain prevention strategies. 


The Human Microbiome Project (HMP)
The Common Fund's Human Microbiome Project (HMP) developed research resources to enable the study of the microbial communities that live in and on our bodies and the roles they play in human health and disease. 
`
const CFDEDocumentationTool = [
	"CFDEDocumentationTool",
	{
		title: "CFDEDocumentationTool",
		description: "Returns information of all the DCCs and Centers involved in CFDE as well as relevant documentation",
		inputSchema: {},
		outputSchema: {
			dccs: z.array(
				z.object({
					label: z.string().describe("long name of the dcc or center"),
					short_label: z.string().describe("short name of the dcc or center"),
					portal_page: z.string().describe("The corresponding page of the dcc or center in CFDE Workbench"),
					description: z.string().describe("description of the dcc or center"),
					homepage: z.string().nullable().describe("homepage of the dcc or center"),
					cf_site: z.string().nullable().optional().describe("Page of the dcc in the CommonFund (https://commonfund.nih.gov/)"),
				})).describe("List of Data Coordination Centers (DCCs)"),
			centers: z.array(
				z.object({
					label: z.string().describe("long name of the dcc or center"),
					short_label: z.string().describe("short name of the dcc or center"),
					portal_page: z.string().describe("The corresponding page of the dcc or center in CFDE Workbench"),
					description: z.string().nullable().describe("description of the dcc or center"),
					homepage: z.string().nullable().describe("homepage of the dcc or center"),
				})).describe("List of Centers involved in CFDE"),
			documentation: z.string().describe("Documentation about the different data, standards, and terminologies used in CFDE")
		}
	},
	async () => {
		const results: {[key:string]: Array<{[key: string]: string | null}> | string} = {}
		results["documentation"] = cfde_info + docs
		const dccs = await prisma.dccs.findMany({
				where: {
					active: true
				}
			})
		if (dccs.length > 0) {
			results["dccs"] = dccs.map((dcc)=>({
				label: dcc.label,
				short_label: dcc.short_label,
				portal_page: `https://info.cfde.cloud/dcc/${dcc.short_label}`,
				description: dcc.description,
				homepage: dcc.homepage,
				cf_site: dcc.cf_site,
			}))
		}
		const centers = await prisma.centers.findMany({
				where: {
					active: true,
					short_label: {
						not: "centers"
					}
				}
			})
		if (centers.length > 0) {
			results["centers"] = centers.map((center)=>({
				label: center.label,
				short_label: center.short_label,
				portal_page: `https://info.cfde.cloud/center/${center.short_label}`,
				description: center.description,
				homepage: center.homepage,
			}))
		}
		return {
			content: [
				{
					type: "text",
					text: JSON.stringify(results)
				}
			],
			structuredContent: results
		}
	}
]

export default CFDEDocumentationTool;